### PR TITLE
GH-20213: [C++] Implement cast to/from halffloat

### PR DESCRIPTION
**Pull request by ClifHouck**
### What changes are included in this PR?

This PR implements casting to and from float16 types using the vendored float16 library included in arrow at `cpp/arrrow/util/float16.*`.

### Are these changes tested?

Unit tests are included in this PR.

### Are there any user-facing changes?

In that casts to and from float16 will now work, yes.

* Closes: #20213

### TODO

- [x] Add casts to/from float64.
- [x] String <-> float16 casts.
- [x] Integer <-> float16 casts.
- [x] Tests.
- [x] Update https://github.com/apache/arrow/blob/main/docs/source/status.rst about half float.
- [x] Rebase.
- [x] Run clang format over this PR.
* GitHub Issue: #20213

### CODE REVIEW

**Comment by ClifHouck** [comment]
This still needs work, but I would greatly appreciate someone who is more familiar with the arrow code-base to comment on whether my approach is sane or not.

File: cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc [hunk 1]
```diff
@@ -833,9 +833,14 @@ std::vector<std::shared_ptr<CastFunction>> GetNumericCasts() {
   auto cast_half_float =
       std::make_shared<CastFunction>("cast_half_float", Type::HALF_FLOAT);
   AddCommonCasts(Type::HALF_FLOAT, float16(), cast_half_float.get());
+  DCHECK_OK(cast_half_float.get()->AddKernel(Type::FLOAT,
+      {InputType(Type::FLOAT)}, float16(), CastFloatingToFloating));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc:837]: You can probably fold this into `GetCastToFloating` and `GetCastToInteger`?

**Comment by pitrou** [comment]
Thanks for posting this. The approach is sane, I posted some comments.

You will have to decide whether you want to tackle casting to string, or prefer leaving that for later/another contributor.

File: cpp/src/arrow/compute/kernels/scalar_cast_internal.cc [hunk 2]
```diff
@@ -47,6 +49,36 @@ struct CastPrimitive {
   }
 };
 
+template<> struct CastPrimitive<HalfFloatType, FloatType, enable_if_t<true>> {
```
[the remark this answers is not in the document]
Comment (ClifHouck) [reply cpp/src/arrow/compute/kernels/scalar_cast_internal.cc]: I used this in a slightly modified form in my latest commit, thank you for the suggestion.

File: cpp/src/arrow/compute/kernels/scalar_cast_internal.cc [hunk 3]
```diff
@@ -47,6 +49,36 @@ struct CastPrimitive {
...
+            float val = *in_values;
+            uint16_t converted = Float16(val).bits();
+            std::memcpy(out_values, static_cast<void*>(&converted), sizeof(uint16_t));
+            out_values++;
+            in_values++;
+        }
+    }
+};
+
+template<> struct CastPrimitive<FloatType, HalfFloatType, enable_if_t<true>> {
```
[the remark this answers is not in the document]
Comment (ClifHouck) [reply cpp/src/arrow/compute/kernels/scalar_cast_internal.cc]: I believe you and I attempted this:
```cpp
template<typename OutType>                                                                                                                                                                          
  struct CastPrimitive<OutType, HalfFloatType, enable_if_floating_point<typename OutType::c_type>> {                                                                                                  
      static void Exec(const ArraySpan& arr, ArraySpan* out) {                                                                                                                                        
          using OutT = typename OutType::c_type;                                                                                                                                                      
          const uint16_t* in_values = arr.GetValues<uint16_t>(1);                                                                                                                                     
          OutT* out_values = out->GetValues<OutT>(1);                                                                                                                                                 
          for (int64_t i = 0; i < arr.length; ++i) {                                                                                                                                                  
              *out_values++ = static_cast<OutT>(Float16::FromBits(*in_values++));                                                                                                                     
          }                                                                                                                                                                                           
      }                                                                                                                                                                                               
  };       
```
But it doesn't seem to call
 ```cpp 
explicit Float16::operator double();
``` 
or float. And I don't see how I can use
```cpp
Float16::ToDouble();
Float16::ToFloat();
```
In a templated version. Maybe you can give me a hint? My C++ template-fu is not great.

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_cast_internal.cc]: Perhaps try to replace:
```c++
              *out_values++ = static_cast<OutT>(Float16::FromBits(*in_values++));                
```
with
```c++                                                                                                 
              *out_values++ = OutT{Float16::FromBits(*in_values++)};
```

File: cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc (hunk 1 above)
Comment (ClifHouck) [reply cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc:837]: I'm not sure I follow. I folded converting half-float to float32/64 into `GetCastToFloating`, but I'm not sure how to integrate float32/64 to half-float into `GetCastToFloating`.

File: cpp/src/arrow/compute/kernels/scalar_cast_internal.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_cast_internal.cc]: I'm a bit surprised. I'll take a look when I find time.

**Comment by ClifHouck** [comment]
Please correct me if I'm wrong, but it seems like float32/float64 cast to string is ultimately handled by this vendored library `cpp/src/arrow/vendored/double-conversion`, source repo here: https://github.com/google/double-conversion .

There doesn't seem to be a direct way to cast from a half float ultimately represented by a `uint16_t` to a string, but I wonder if it's acceptable to cast to float first and then use the existing float to string cast? or is that unacceptable overhead?

**Comment by pitrou** [comment]
> There doesn't seem to be a direct way to cast from a half float ultimately represented by a `uint16_t` to a string, but I wonder if it's acceptable to cast to float first and then use the existing float to string cast? or is that unacceptable overhead?

Yes, it's acceptable. The only annoyance would be the superfluous precision, but that's not a blocking issue.

**Comment by pitrou** [comment]
By the way, you'll have to implement unit tests at some point (in C++).

**Comment by ianmcook** [comment]
@ClifHouck thanks for working on this!

Small bit of housekeeping: 

There is a note in [`docs/source/status.rst`](https://github.com/apache/arrow/blob/main/docs/source/status.rst) that says:
>(1) Casting to/from Float16 in C++ is not supported.

Could you please push a commit here that removes that note? Looks like you will need to rebase first because there have been some changes in `status.rst` since you began work on this PR.

**Comment by ClifHouck** [comment]
So one question I currently have is, should I specialize `Datum` on `HalfFloatType` (or similar)? Right now tests are failing when trying to compare expected to actual output from doing casts from, say, int to float16.

**Comment by pitrou** [comment]
> So one question I currently have is, should I specialize `Datum` on `HalfFloatType` (or similar)?

Which method would you specialize?

**Comment by ClifHouck** [comment]
> > So one question I currently have is, should I specialize `Datum` on `HalfFloatType` (or similar)?
> 
> Which method would you specialize?

After reading more code, more likely I need to specialize or modify `CompareFloating`, `VisitFloatingEquality`. But I'm not certain.

**Comment by pitrou** [comment]
> After reading more code, more likely I need to specialize or modify `CompareFloating`, `VisitFloatingEquality`, etc. But I'm not certain.

If you're struggling on this, someone else can take a look and help. Perhaps @benibus or @felipecrv ?
(or I can get to it later when I have time)

**Comment by ClifHouck** [comment]
I'm currently stuck on getting a new `FloatingToFloating` test to pass. AFAICT the other existing tests now pass with the new float16 casts.

**Comment by ClifHouck** [comment]
I think I've fixed things with my latest commit. Turned out a specialization of `CastPrimitive` was not being called as I expected, and instead the generic template version was being used.

**Comment by ClifHouck** [comment]
I added a few more TODOs but I think this is very close to being ready for review.

File: cpp/src/arrow/compute/kernels/scalar_cast_internal.cc [hunk 4]
```diff
@@ -56,6 +93,36 @@ struct CastPrimitive<OutType, InType, enable_if_t<std::is_same<OutType, InType>:
...
+          *out_values++ = Float16(temp).bits();
+        }
+   }
+};
+
+// Cast half float to int
+template<typename OutType>
+struct CastPrimitive<OutType, HalfFloatType, enable_if_integer<OutType>> {
+   static void Exec(const ArraySpan& arr, ArraySpan* out) {
+        std::cout << "HalfFloat to Int CastPrimitive()!\n";
```
Comment (ClifHouck) [inline cpp/src/arrow/compute/kernels/scalar_cast_internal.cc:114]: Need to remove this.

File: cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc [hunk 5]
```diff
@@ -151,6 +174,10 @@ Status CheckFloatToIntTruncation(const ExecValue& input, const ExecResult& outpu
       return CheckFloatToIntTruncationImpl<FloatType>(input.array, *output.array_span());
     case Type::DOUBLE:
       return CheckFloatToIntTruncationImpl<DoubleType>(input.array, *output.array_span());
+    // FIXME: Is this OK?  It seems like the underlying check may not operator correctly
```
Comment (ClifHouck) [inline cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc:177]: Need to remove this.

**Comment by ClifHouck** [comment]
I think this is close enough for a round of reviews. There's some iffy-ness around `ConvertNumber` for `HalfFloatType` and the test situation, but hopefully can be ironed out with reviews from people with more arrow experience.

**Comment by pitrou** [comment]
For the Docker-based builds (i.e. all Linux builds), you can take a look at https://arrow.apache.org/docs/developers/continuous_integration/docker.html

**Comment by ClifHouck** [comment]
I believe the last commit resolves the test failures.

**Comment by ClifHouck** [comment]
It seems like the 2 R related checks are failing due to something about duckdb: https://github.com/apache/arrow/actions/runs/8366814875/job/22912629785?pr=40067#step:7:146

Not sure if this is problem caused by the changes in this PR or if this is something unrelated.

**Comment by ClifHouck** [comment]
Yep, looks like I'm running into this: https://github.com/apache/arrow/issues/40702

**Comment by pitrou** [comment]
Yes, those are unrelated, sorry.

**Comment by ClifHouck** [comment]
Rebased on latest main with duckdb fix

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 6]
```diff
@@ -2228,6 +2237,10 @@ TEST(Cast, FloatingToString) {
     CheckCast(
         ArrayFromJSON(float64(), "[0.0, -0.0, 1.5, -Inf, Inf, NaN, null]"),
         ArrayFromJSON(string_type, R"(["0", "-0", "1.5", "-inf", "inf", "nan", null])"));
+
+    CheckCast(
+        ArrayFromJSON(float16(), "[0.0, -0.0, 1.5, -Inf, Inf, NaN, null]"),
+        ArrayFromJSON(string_type, R"(["0", "-0", "1.5", "-inf", "inf", "nan", null])"));
```
Comment (kou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:2243]: How about using `float16()` -> `float32()` -> `float64()` order like other places?

```cpp
    CheckCast(
        ArrayFromJSON(float16(), "[0.0, -0.0, 1.5, -Inf, Inf, NaN, null]"),
        ArrayFromJSON(string_type, R"(["0", "-0", "1.5", "-inf", "inf", "nan", null])"));

    CheckCast(
        ArrayFromJSON(float32(), "[0.0, -0.0, 1.5, -Inf, Inf, NaN, null]"),
        ArrayFromJSON(string_type, R"(["0", "-0", "1.5", "-inf", "inf", "nan", null])"));

    CheckCast(
        ArrayFromJSON(float64(), "[0.0, -0.0, 1.5, -Inf, Inf, NaN, null]"),
        ArrayFromJSON(string_type, R"(["0", "-0", "1.5", "-inf", "inf", "nan", null])"));
```

Or we may want to use `for (auto float_type : {float16(), float32(), float64()})` style:

```cpp
  for (auto float_type : {float16(), float32(), float64()}) {
    for (auto string_type : {utf8(), large_utf8()}) {
      CheckCast(
          ArrayFromJSON(float_type, "[0.0, -0.0, 1.5, -Inf, Inf, NaN, null]"),
          ArrayFromJSON(string_type, R"(["0", "-0", "1.5", "-inf", "inf", "nan", null])"));
    }
  }
```

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 6 above)
Comment (ClifHouck) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: Refactored to match this, thanks for the suggestion.

**Comment by ClifHouck** [comment]
Looks like CI issues again.

**Comment by kou** [comment]
They are caused by #40744. This PR isn't related.

**Comment by ClifHouck** [comment]
Is there any way to re-kick the failing test? It looks like a transient CI issue.

**Comment by kou** [comment]
Could you rebase on main?
The failure was fixed by #40794.

**Review by kou — COMMENTED** [review]
I'll merge in a few days if nobody objects it.

**Comment by ClifHouck** [comment]
> Could you rebase on main? The failure was fixed by #40794.

Done.

**Comment by ianmcook** [comment]
Any objections to @kou merging this?

**Comment by kou** [comment]
Oh, sorry. I missed this. I'll merge this.

File: cpp/src/arrow/ipc/json_simple_test.cc [hunk 7]
```diff
@@ -185,6 +189,21 @@ class TestIntegers : public ::testing::Test {
 
 TYPED_TEST_SUITE_P(TestIntegers);
 
+template <typename DataType>
+std::vector<typename DataType::c_type> TestIntegersMutateIfNeeded(
+    std::vector<typename DataType::c_type> data) {
+  return data;
+}
+
+// TODO: This works, but is it the right way to do this?
```
Comment (pitrou) [inline cpp/src/arrow/ipc/json_simple_test.cc:198]: I'm curious, what is this function meant for?

File: cpp/src/arrow/ipc/json_simple_test.cc [hunk 8]
```diff
@@ -269,7 +289,12 @@ INSTANTIATE_TYPED_TEST_SUITE_P(TestUInt8, TestIntegers, UInt8Type);
 INSTANTIATE_TYPED_TEST_SUITE_P(TestUInt16, TestIntegers, UInt16Type);
 INSTANTIATE_TYPED_TEST_SUITE_P(TestUInt32, TestIntegers, UInt32Type);
 INSTANTIATE_TYPED_TEST_SUITE_P(TestUInt64, TestIntegers, UInt64Type);
-INSTANTIATE_TYPED_TEST_SUITE_P(TestHalfFloat, TestIntegers, HalfFloatType);
+// FIXME: I understand that HalfFloatType is backed by a uint16_t, but does it
+// make sense to run this test over it?
```
Comment (pitrou) [inline cpp/src/arrow/ipc/json_simple_test.cc:293]: No, I don't think so.

File: cpp/src/arrow/util/value_parsing.h [hunk 9]
```diff
@@ -163,6 +166,20 @@ struct StringConverter<DoubleType> {
   const char decimal_point;
 };
 
+template <>
+struct StringConverter<HalfFloatType> {
```
Comment (pitrou) [inline cpp/src/arrow/util/value_parsing.h:170]: Should have added some tests for value parsing and formatting. @kou

**Comment by pitrou** [comment]
Oops, sorry that I couldn't give this a review earlier.

File: cpp/src/arrow/util/value_parsing.h (hunk 9 above)
Comment (kou) [reply cpp/src/arrow/util/value_parsing.h:170]: Ah, I thought tests added to `scalar_cast_test.cc` cover this.
I missed that we have `value_parsing_test.cc`.

@ClifHouck Could you add tests for this to `value_parsing_test.cc` as a follow-up task?

Comment (pitrou) [reply cpp/src/arrow/util/value_parsing.h:170]: And also `formatting_util_test.cc`

**Comment by ClifHouck** [comment]
I'll address any additional code-related changes needed here in a follow-up PR.

File: cpp/src/arrow/util/value_parsing.h (hunk 9 above)
Comment (ClifHouck) [reply cpp/src/arrow/util/value_parsing.h:170]: I'll take a look and add tests.

File: cpp/src/arrow/ipc/json_simple_test.cc (hunk 7 above)
Comment (ClifHouck) [reply cpp/src/arrow/ipc/json_simple_test.cc:198]: I should remove this since we probably don't need to run this test on `HalfFloatType`.

**Pull request merged**